pkg/operator/ceph/cluster/cluster_cleanup.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster to manage a Ceph cluster.
+package cluster
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookv1 "github.com/rook/rook/pkg/apis/rook.io/v1"
+	"github.com/rook/rook/pkg/operator/ceph/controller"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	batch "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	appName              = "cluserCleanUp"
+	dataDirHostPath      = "ROOK_DATA_DIR_HOST_PATH"
+	cleanDataDirHostPath = "ROOK_CLEAN_DATA_DIR_HOST_PATH"
+)
+
+func (c *ClusterController) startClusterCleanUp(cluster *cephv1.CephCluster, retryInterval time.Duration) {
+	logger.Infof("starting clean up for cluster %q", cluster.Name)
+	err := c.waitForCephDaemonCleanUp(cluster, time.Duration(clusterDeleteRetryInterval)*time.Second)
+	if err != nil {
+		logger.Errorf("failed to wait till ceph daemons are destroyed. %v", err)
+		return
+	}
+	c.startCleanUpJob(cluster)
+}
+
+func (c *ClusterController) startCleanUpJob(cluster *cephv1.CephCluster) {
+	validStorage := cluster.Spec.Storage
+	privileged := true
+
+	// Get the list of all nodes in the cluster.
+	hostnameMap, err := k8sutil.GetNodeHostNames(c.context.Clientset)
+	if err != nil {
+		logger.Infof("error finding hostname. %+v", err)
+		return
+	}
+	for _, hostname := range hostnameMap {
+		storageNode := rookv1.Node{
+			Name: hostname,
+		}
+		validStorage.Nodes = append(validStorage.Nodes, storageNode)
+	}
+
+	validNodes := k8sutil.GetValidNodes(validStorage, c.context.Clientset, cephv1.GetOSDPlacement(cluster.Spec.Placement))
```

Comment:
We need to remove the dataDirHostPath on all the nodes where any Ceph daemon was running, not just the OSDs. All the daemons mount the host path. So before the daemons stop, we will need to query all the rook pods in the cluster namespace, and detect the set of nodes they are running on.